### PR TITLE
[ Kernel ] AWQ Fused MoE

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
SUMMARY:
* Picks up #6259 2761 to support AWQ MoE models via a fused kernel, after the refactor I did in #5970. The kernels for this PR were developed by @chu-tianxiang
* Adds `AWQMoEmethod`, supporting loading AutoAWQ models
* Refactors `FusedMoE.weight_loader`, to enable loading AWQ models, which have transposed weights (input_dim, output_dim) on disk. Fp16 and Fp8 models have share (input_dim, output_dim). This required more complex logic for handling indexing in the TP case and MergedColumn case
* Refactors `expert_params_mapping`, which was overfit to fp16 and fp8 checkpoints. This required renaming the scale parameters in fp8 which to better match the state dicts that we create in autofp8, limiting the amount of remapping we need to do in the model files
* Updates layers to use `fused_topk`/`grouped_topk` and `fused_experts`, rather than calling `fused_moe` directly, such that we can reuse the logic across `fp16`, `fp8`, and `awq`

Latency Benchmarking with 2xA100 80GB:

```
Mixtral Fused MoE with AWQ:
> python benchmarks/benchmark_latency.py --model casperhansen/mixtral-instruct-awq --tensor-parallel-size 2 --input-len 512 --output-len 128 --max-model-len 1024 --quantization awq --batch-size 1
Avg latency: 1.3650233147976298 seconds
10% percentile latency: 1.3638953405432404 seconds
25% percentile latency: 1.3643284866120666 seconds
50% percentile latency: 1.3648834400810301 seconds
75% percentile latency: 1.3656865148805082 seconds
90% percentile latency: 1.3661401799879969 seconds
99% percentile latency: 1.366987234679982 seconds

Mixtral Fp16:
> python benchmarks/benchmark_latency.py --model mistralai/Mixtral-8x7B-Instruct-v0.1 --tensor-parallel-size 2 --input-len 512 --output-len 128 --max-model-len 1024 --batch-size 1
Avg latency: 1.470517885716011 seconds
10% percentile latency: 1.4687837794423104 seconds
25% percentile latency: 1.4693051476497203 seconds
50% percentile latency: 1.4700921354815364 seconds
75% percentile latency: 1.4719352358952165 seconds
90% percentile latency: 1.4728145461529494 seconds
99% percentile latency: 1.4740593627933414 seconds
```

Followup work:
* Deepseek is not supported yet, illegal memory access with `TechxGenus/DeepSeek-V2-Lite-Chat-AWQ `

### CODE REVIEW

**Comment by AlpinDale** [comment]
Thanks for doing this.

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 1]
```diff
@@ -61,66 +58,41 @@ def create_weights(self, layer: torch.nn.Module, num_experts: int,
...
         assert not use_grouped_topk
-        assert num_expert_group is None
-        assert topk_group is None
-        return fused_moe(x, w1, w2, router_logits, top_k, renormalize)
+        return fused_experts(hidden_states=x,
+                             w1=layer.w13_weight,
+                             w2=layer.w2_weight,
+                             topk_weights=topk_weights,
+                             topk_ids=topk_ids,
+                             inplace=True)
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/layer.py:95]: It looks like inplace wasn't true for tpu before?

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 2]
```diff
@@ -210,92 +198,99 @@ def weight_loader(self, param: torch.nn.Parameter,
...
+            return
+
+        expert_data = param.data[expert_id]
+        tp_rank = get_tensor_model_parallel_rank()
+
+        # If transposed, weight is saved as [input_dim, output_dim]
+        # Otherwise, weight is saved as     [output_dim, input_dim]
+        is_transposed = getattr(param, "is_transposed", False)
+        input_dim = 0 if is_transposed else 1
+        output_dim = 1 if is_transposed else 0
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/layer.py:230]: We should read input_dim and output_dim from the layer instead, this is a common pattern in the codebase: https://github.com/vllm-project/vllm/blob/05308891e203329a733bcf29a3452b15b75b5eb4/vllm/model_executor/layers/quantization/fp8.py#L151-L155

File: vllm/model_executor/layers/quantization/awq.py [hunk 3]
```diff
@@ -174,3 +179,117 @@ def apply(self,
...
+                                            2 * intermediate_size //
+                                            self.quant_config.pack_factor,
+                                            dtype=torch.int32),
+                                requires_grad=False)
+        layer.register_parameter("w13_qweight", w13_qweight)
+        set_weight_attrs(
+            w13_qweight, {
+                "packed_dim": 1,
+                "pack_factor": self.quant_config.pack_factor,
+                "is_transposed": True,
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/awq.py:205]: Ditto on replacing is_transposed with explicit input/output_dim

**Review by mgoin — COMMENTED** [review]
This looks to be shaping up well! I appreciate the refactoring of fused_moe. It would be more clear if it was separated out first, but I think we can land it within this PR if needed.

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 4]
```diff
@@ -442,8 +442,8 @@ def fused_experts(hidden_states: torch.Tensor,
     assert hidden_states.shape[1] == w1.shape[2], "Hidden size mismatch"
     assert topk_weights.shape == topk_ids.shape, "topk shape mismatch"
     assert hidden_states.is_contiguous(), "Hidden_states must be contiguous"
-    assert w1.is_contiguous(), "Expert weights1 must be contiguous"
-    assert w2.is_contiguous(), "Expert weights2 must be contiguous"
+    #assert w1.is_contiguous(), "Expert weights1 must be contiguous"
+    #assert w2.is_contiguous(), "Expert weights2 must be contiguous"
```
[the remark this answers is not in the document]
Comment (dsikka) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:446]: Good catch

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 1 above)
Comment (dsikka) [reply vllm/model_executor/layers/fused_moe/layer.py:95]: good catch

Comment (dsikka) [reply vllm/model_executor/layers/fused_moe/layer.py:95]: Actually, seems to using a `fused_moe` function which doesnt have the inplace flag? 
https://github.com/vllm-project/vllm/blob/16a1cc9bb2b4bba82d78f329e5a89b44a5523ac8/vllm/model_executor/layers/fused_moe/moe_pallas.py#L6

**Comment by dsikka** [comment]
@mgoin can't resolve comments but addressed all but one comment

File: vllm/model_executor/layers/fused_moe/fused_moe_awq.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                             topk_weights, topk_ids)
+
+    (sorted_token_ids, expert_ids,
+     num_tokens_post_padded) = moe_align_block_size(topk_ids, 16, w1.shape[0])
+
+    x = hidden_states.view(hidden_states.shape[0], 1, *hidden_states.shape[1:])
+
+    gate_up = ops.awq_fused_moe(x, w1, w1_scales, w1_qzeros, topk_weights,
+                                sorted_token_ids, expert_ids,
+                                num_tokens_post_padded, False, pack_factor)
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/fused_moe_awq.py:62]: It would be good to use named args for the last few args at least since a floating "False" is unclear

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 1 above)
Comment (mgoin) [reply vllm/model_executor/layers/fused_moe/layer.py:95]: Yeah this isn't correct at the moment, it needs to keep using the pallas fused_moe function. I think it is fine to revert/specialize here like before

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 6]
```diff
@@ -210,92 +198,99 @@ def weight_loader(self, param: torch.nn.Parameter,
...
-            elif shard_id == 2:
-                param_data[expert_id, shard_size:2 *
-                           shard_size, :] = loaded_weight[shard, :]
-            # w2, down_proj case: Load into only shard of w2.
-            elif shard_id == 1:
-                param_data[expert_id, :, :] = loaded_weight[:, shard]
-            else:
-                raise ValueError(
-                    f"Shard id must be in [0,1,2] but got {shard_id}")
+            raise ValueError
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/layer.py:257]: Can you keep an error message with the shard_id printed?

File: vllm/model_executor/models/deepseek_v2.py [hunk 7]
```diff
@@ -115,7 +115,7 @@ def __init__(
                                 hidden_size=config.hidden_size,
                                 intermediate_size=config.moe_intermediate_size,
                                 reduce_results=False,
-                                renormalize=config.norm_topk_prob,
+                                renormalize=True,
```
Comment (mgoin) [inline vllm/model_executor/models/deepseek_v2.py:118]: Why is this changed from the config?

Comment (dsikka) [reply vllm/model_executor/models/deepseek_v2.py:118]: Ah this was to fix the memory issue for awq. I realized this was the flag that was triggering the illegal memory issue so set it to True to prevent the error. Reverting

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 1 above)
Comment (dsikka) [reply vllm/model_executor/layers/fused_moe/layer.py]: Updated

**Comment by dsikka** [comment]
Latency Benchmark with `mixtral`:

```
Avg latency: 1.502528661613663 seconds
10% percentile latency: 1.4870146293193103 seconds
25% percentile latency: 1.4933206527493894 seconds
50% percentile latency: 1.5005597593262792 seconds
75% percentile latency: 1.5073543908074498 seconds
90% percentile latency: 1.5178802264854312 seconds
99% percentile latency: 1.5486325215920806 seconds
```

**Comment by dsikka** [comment]
Latency Benchmarking with Two 82 GB A100s:

```
Mixtral Fused MoE with AWQ:
Avg latency: 1.3650233147976298 seconds
10% percentile latency: 1.3638953405432404 seconds
25% percentile latency: 1.3643284866120666 seconds
50% percentile latency: 1.3648834400810301 seconds
75% percentile latency: 1.3656865148805082 seconds
90% percentile latency: 1.3661401799879969 seconds
99% percentile latency: 1.366987234679982 seconds

Mixtral Fp16:
Avg latency: 1.470517885716011 seconds
10% percentile latency: 1.4687837794423104 seconds
25% percentile latency: 1.4693051476497203 seconds
50% percentile latency: 1.4700921354815364 seconds
75% percentile latency: 1.4719352358952165 seconds
90% percentile latency: 1.4728145461529494 seconds
99% percentile latency: 1.4740593627933414 seconds
```

File: vllm/model_executor/layers/fused_moe/fused_moe_awq.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+import torch
+
+from vllm import _custom_ops as ops
+from vllm.logger import init_logger
+
+from .fused_moe import fused_experts, moe_align_block_size
+
+logger = init_logger(__name__)
+
+NAIVE_THRESHOLD = 1024
```
[the remark this answers is not in the document]
Comment (dsikka) [reply vllm/model_executor/layers/fused_moe/fused_moe_awq.py:11]: @robertgshaw2-neuralmagic do we know why this is 1024 specifically?

**Review by mgoin — APPROVED** [review]
LGTM! We should add a model test, but considering there isn't a small mixtral awq to test, maybe just including a [lm-eval large test](https://github.com/vllm-project/vllm/blob/746709642c81aa22926765aef67e086a15aef076/.buildkite/lm-eval-harness/configs/models-large.txt#L3) would be sufficient.

**Comment by dsikka** [comment]
Note: Splitting this PR into two separate PRs.
PR 1/2: https://github.com/vllm-project/vllm/pull/7334
